File: generate_role_hashes.py
import urllib.request
from PIL import Image
import io

# URLs for the role icons in Wild Rift
roles = {
    "TOP": "https://raw.communitydragon.org/latest/plugins/rcp-fe-lol-static-assets/global/default/images/icon-position-top.png",
    "JUNGLE": "https://raw.communitydragon.org/latest/plugins/rcp-fe-lol-static-assets/global/default/images/icon-position-jungle.png",
    "MID": "https://raw.communitydragon.org/latest/plugins/rcp-fe-lol-static-assets/global/default/images/icon-position-middle.png",
    "ADC": "https://raw.communitydragon.org/latest/plugins/rcp-fe-lol-static-assets/global/default/images/icon-position-bottom.png",
    "SUPPORT": "https://raw.communitydragon.org/latest/plugins/rcp-fe-lol-static-assets/global/default/images/icon-position-utility.png"
}

def ahash(image):
    image = image.resize((8, 8), Image.Resampling.LANCZOS).convert('L')
    pixels = list(image.getdata())
    avg = sum(pixels) / len(pixels)
    hash_val = 0
    for i, p in enumerate(pixels):
        if p >= avg:
            hash_val |= (1 << (63 - i))
    return hash_val

hashes = {}
for role, url in roles.items():
    try:
        req = urllib.request.Request(url, headers={'User-Agent': 'Mozilla/5.0'})
        with urllib.request.urlopen(req) as response:
            img_data = response.read()
        img = Image.open(io.BytesIO(img_data))
        h = ahash(img)
        hashes[role] = h
        print(f"Downloaded {role}: {h}")
    except Exception as e:
        print(f"Failed {role}: {e}")

with open('app/src/main/java/com/example/util/RoleHashes.kt', 'w') as f:
    f.write("package com.example.util\n\n")
    f.write("import com.example.model.LaneRole\n\n")
    f.write("object RoleHashes {\n")
    f.write("    val map = mapOf<LaneRole, Long>(\n")
    for k, v in hashes.items():
        f.write(f'        LaneRole.{k} to {v}L,\n')
    f.write("    )\n")
    f.write("}\n")

File: generate_role_hashes.java
import java.io.*;
import java.net.URL;
import java.awt.Graphics2D;
import java.awt.Image;
import java.awt.image.BufferedImage;
import javax.imageio.ImageIO;
import java.nio.file.Files;
import java.nio.file.Paths;

public class generate_role_hashes {
    public static void main(String[] args) throws Exception {
        String[][] roles = {
            {"TOP", "https://raw.communitydragon.org/latest/plugins/rcp-fe-lol-static-assets/global/default/images/icon-position-top.png"},
            {"JUNGLE", "https://raw.communitydragon.org/latest/plugins/rcp-fe-lol-static-assets/global/default/images/icon-position-jungle.png"},
            {"MID", "https://raw.communitydragon.org/latest/plugins/rcp-fe-lol-static-assets/global/default/images/icon-position-middle.png"},
            {"ADC", "https://raw.communitydragon.org/latest/plugins/rcp-fe-lol-static-assets/global/default/images/icon-position-bottom.png"},
            {"SUPPORT", "https://raw.communitydragon.org/latest/plugins/rcp-fe-lol-static-assets/global/default/images/icon-position-utility.png"}
        };
        
        StringBuilder out = new StringBuilder("package com.example.util\n\nimport com.example.model.LaneRole\n\nobject RoleHashes {\n    val map = mapOf<LaneRole, Long>(\n");

        for (String[] role : roles) {
            try {
                URL url = new URL(role[1]);
                java.net.HttpURLConnection con = (java.net.HttpURLConnection)url.openConnection();
                con.setRequestProperty("User-Agent", "Mozilla/5.0");
                BufferedImage img = ImageIO.read(con.getInputStream());
                
                BufferedImage resized = new BufferedImage(8, 8, BufferedImage.TYPE_BYTE_GRAY);
                Graphics2D g2 = resized.createGraphics();
                g2.drawImage(img, 0, 0, 8, 8, null);
                g2.dispose();
                
                long total = 0;
                int[] pixels = new int[64];
                int idx = 0;
                for (int y = 0; y < 8; y++) {
                    for (int x = 0; x < 8; x++) {
                        int p = resized.getRGB(x, y) & 0xFF;
                        pixels[idx++] = p;
                        total += p;
                    }
                }
                
                long avg = total / 64;
                long hash = 0;
                for (int j = 0; j < 64; j++) {
                    if (pixels[j] >= avg) {
                        hash |= (1L << (63 - j));
                    }
                }
                out.append(String.format("        LaneRole.%s to %dL,\n", role[0], hash));
                System.out.println("Downloaded " + role[0] + ": " + hash);
            } catch (Exception e) {
                System.out.println("Fail " + role[0] + ": " + e.getMessage());
            }
        }
        
        out.append("    )\n}\n");
        Files.write(Paths.get("app/src/main/java/com/example/util/RoleHashes.kt"), out.toString().getBytes());
        System.out.println("Done Writing");
    }
}
